Merge in ERC1155 Smart Contracts & Tests

pynchmeister opened the pull request:
Wrote and tested ERC1155 Contracts and their various functionality. Navigate into ```'../hardhat/contracts/token/ERC1155'``` and refer to the package.json to build files and run tests.

## Future Work

* create an identical PR in the hardhat-etc repo

kimanikelly reviewed (commented):
- The ```ERC1155``` folder should be separated from the ``` hardhat/contracts/token``` folder to prevent uncertainties between legacy contracts and new contracts.
 - The ```ERC1155``` folder should only contain contracts to persist the same structure in ``` hardhat/contracts```.
 - ```hardhat/contracts/token/ERC1155/hardhat.config.ts``` Is a duplicate of the original ```hardhat.config.ts``` file located in the root of the ```hardhat``` folder. This is creating a second hardhat environment inside the ``` hardhat/contracts``` folder.
 - Multiple package.json files nested inside the ```ERC1155``` folder.
 - The unit tests for the contracts should be removed from the ```ERC1155``` folder and moved to the ```test``` folder located in the root of ```hardhat```
 - The ERC1155 contracts cannot be compiled from the root of the ```hardhat``` folder like the other contracts can.
 - The ERC1155 tests cannot be executed from the root of the ```hardhat``` folder like the other tests can.
 - If the these contracts needed to be deployed with the current folder structure it would have to be done within the ```hardhat/contracts``` folder because of the second hardhat environment.

pynchmeister commented:
Fair critique. I will work to fix this, any guidance is appreciated.

pynchmeister commented:
In closing this issue I refactored the tests in Solidity and switched to Foundry instead of Hardhat, the code can be read here: https://github.com/zapproject/foundry
